Rootcheck not generating findings (only start/end events stored)

## Description

Rootcheck is currently not generating or persisting any findings in the agent databases. Only the start and end scan events are stored in the `pm_event` table:

- `Starting rootcheck scan.`
- `Ending rootcheck scan.`

As a result, all `/rootcheck/{agent_id}` API calls return empty results, causing related AITs to fail.

```console
FAILED test_rootcheck_endpoints.tavern.yaml::GET /rootcheck/001
FAILED test_rootcheck_endpoints.tavern.yaml::DELETE /rootcheck/{agent_id}
```

---

## Impact

Several AITs related to rootcheck endpoints are failing because they expect real findings in `pm_event`, but the database only contains start/end scan events.

Example:

```json
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "No rootcheck information was returned",
  "error": 0
}
```

However, `/rootcheck/{agent_id}/last_scan` works correctly:

```json
{
  "data": {
    "affected_items": [
      {
        "start": "2026-03-18T12:37:24Z",
        "end": "2026-03-18T12:37:33Z"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "Last rootcheck scan of the agent was returned",
  "error": 0
}
```

This confirms that rootcheck scans are executed but no findings are stored.

---

## Database inspection

Direct inspection of multiple agent databases shows that no findings are present.

Example (`001.db`):

```sql
SELECT COUNT(*) FROM pm_event;
-- Result: 2

SELECT id, log FROM pm_event;
-- 1 | Starting rootcheck scan.
-- 2 | Ending rootcheck scan.

SELECT id, log FROM pm_event
WHERE log NOT IN ('Starting rootcheck scan.', 'Ending rootcheck scan.');
-- (empty)
```

Same result observed in:

- `003.db`
- `005.db`

---

## Rootcheck runtime evidence

Agent logs show repeated errors:

```text
root@wazuh-agent1:/# grep -n -iE "rootcheck: ERROR|rootcheck: WARNING" /var/ossec/logs/ossec.log
11243:2026/03/18 12:32:26 rootcheck: ERROR: 'ps' not found.
11980:2026/03/18 12:32:46 rootcheck: ERROR: 'ps' not found.
23414:2026/03/18 12:33:05 rootcheck: ERROR: 'ps' not found.
34678:2026/03/18 12:37:24 rootcheck: ERROR: 'ps' not found.
```

---

## Verification of ps binary

Inside the agent, `ps` is present and works correctly:

```bash
root@wazuh-agent1:/# ps
    PID TTY          TIME CMD
   6927 pts/0    00:00:00 bash
   6939 pts/0    00:00:00 ps
```

And the binary exists:

```bash
/bin/ps exists: True
/usr/bin/ps exists: True
```

This indicates that:

- `ps` is installed and executable
- but rootcheck is not detecting it correctly during runtime

---

## Possible cause

This behavior may be related to the file detection logic used by rootcheck (`is_file()`), which has a known issue:

https://github.com/wazuh/wazuh/issues/15168

If `ps` is not properly detected:

- Process-related checks may not execute
- No `ALERT_ROOTKIT_FOUND` events are generated
- No findings are inserted into `pm_event`

---

## Summary

- Rootcheck scans are executed (start/end events present)
- No findings are generated or persisted
- API endpoints behave correctly based on DB contents
- `ps` is installed and works, but rootcheck does not detect it


Artifacts: [artifacts.zip](https://github.com/user-attachments/files/26136039/artifacts.zip)
Related issue: https://github.com/wazuh/wazuh/issues/34903

--- a/src/rootcheck/common.c
+++ b/src/rootcheck/common.c
@@ -445,37 +445,35 @@ int isfile_ondir(const char *file, const char *dir)
  */
 int is_file(char *file_name)
 {
-    struct  stat statbuf;
-    int     ret = 0;
-    FILE    *fp = NULL;
-    DIR     *dp = NULL;
-
     if (!file_name) {
         mtdebug2(ARGV0, "RK: Invalid file name: NULL!");
-        return ret;
+        return 0;
     }
 
-    dp = wopendir(file_name);
+    DIR* dp = wopendir(file_name);
     if (dp) {
         closedir(dp);
-        ret = 1;
+        return 1;
     }
 
-    /* Trying other calls */
-    if ((w_stat(file_name, &statbuf) < 0) &&
+    struct  stat statbuf;
+    if (w_stat(file_name, &statbuf) == 0) {
+        return 1;
+    }
+
 #ifndef WIN32
-            (waccess(file_name, F_OK) < 0) &&
+    if (waccess(file_name, F_OK) == 0) {
+        return 1;
+    }
 #endif
-            ((fp = wfopen(file_name, "r")) == NULL)) {
-        return ret;
-    }
 
+    FILE* fp = wfopen(file_name, "r");
     if (fp) {
-        ret = 1;
         fclose(fp);
+        return 1;
     }
 
-    return ret;
+    return 0;
 }
 
 /* Delete the process list */

--- a/src/unit_tests/syscheckd/CMakeLists.txt
+++ b/src/unit_tests/syscheckd/CMakeLists.txt
@@ -359,6 +359,23 @@ if(NOT ${TARGET} STREQUAL "winagent")
   target_include_directories(test_check_rc_sys PRIVATE ${SRC_FOLDER}/rootcheck)
   target_link_libraries(test_check_rc_sys SYSCHECK_O ${TEST_DEPS} "${CHECK_RC_SYS_FLAGS}")
   add_test(NAME test_check_rc_sys COMMAND test_check_rc_sys)
+
+  set(COMMON_FLAGS "-Wl,--wrap,wopendir \
+                    -Wl,--wrap,closedir \
+                    -Wl,--wrap,w_stat \
+                    -Wl,--wrap,waccess \
+                    -Wl,--wrap,wfopen \
+                    -Wl,--wrap,fclose \
+                    -Wl,--wrap,OS_Regex \
+                    -Wl,--wrap,OS_Match2 \
+                    -Wl,--wrap,OSList_GetFirstNode \
+                    -Wl,--wrap,OSList_GetNextNode \
+                    ${DEBUG_OP_WRAPPERS}")
+
+  add_executable(test_common test_common.c)
+  target_include_directories(test_common PRIVATE ${SRC_FOLDER}/rootcheck)
+  target_link_libraries(test_common SYSCHECK_O ${TEST_DEPS} "${COMMON_FLAGS}")
+  add_test(NAME test_common COMMAND test_common)
 endif()
 
 # create_db.c tests

--- a/src/unit_tests/syscheckd/test_common.c
+++ b/src/unit_tests/syscheckd/test_common.c
@@ -1,0 +1,220 @@
+/* Copyright (C) 2026, Wazuh Inc.
+ * All rights reserved.
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation
+ */
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+#include <dirent.h>
+#include <sys/stat.h>
+
+#include "shared.h"
+#include "rootcheck.h"
+
+/* Required globals declared extern in rootcheck.h */
+rkconfig rootcheck;
+int rk_sys_count;
+char **rk_sys_file;
+char **rk_sys_name;
+int test_mode = 1;
+
+/* ===================================================================
+ * Wrappers
+ * =================================================================== */
+
+DIR *__wrap_wopendir(const char *name)
+{
+    check_expected(name);
+    return mock_type(DIR *);
+}
+
+int __wrap_closedir(DIR *dirp)
+{
+    (void)dirp;
+    return 0;
+}
+
+int __wrap_w_stat(const char *pathname, struct stat *buf)
+{
+    check_expected(pathname);
+    (void)buf;
+    return mock_type(int);
+}
+
+int __wrap_waccess(const char *path, int mode)
+{
+    check_expected(path);
+    (void)mode;
+    return mock_type(int);
+}
+
+FILE *__wrap_wfopen(const char *path, const char *mode)
+{
+    check_expected(path);
+    (void)mode;
+    return mock_type(FILE *);
+}
+
+int __wrap_fclose(FILE *stream)
+{
+    (void)stream;
+    return 0;
+}
+
+void __wrap__mtdebug1(__attribute__((unused)) const char *tag,
+                      __attribute__((unused)) const char *file,
+                      __attribute__((unused)) int line,
+                      __attribute__((unused)) const char *func,
+                      __attribute__((unused)) const char *msg, ...) {}
+
+void __wrap__mtdebug2(__attribute__((unused)) const char *tag,
+                      __attribute__((unused)) const char *file,
+                      __attribute__((unused)) int line,
+                      __attribute__((unused)) const char *func,
+                      __attribute__((unused)) const char *msg, ...) {}
+
+void __wrap__merror_exit(__attribute__((unused)) const char *tag,
+                         __attribute__((unused)) const char *file,
+                         __attribute__((unused)) int line,
+                         __attribute__((unused)) const char *func,
+                         __attribute__((unused)) const char *msg, ...) {}
+
+int __wrap_OS_Regex(__attribute__((unused)) const char *pattern,
+                    __attribute__((unused)) const char *str)
+{
+    return 0;
+}
+
+int __wrap_OS_Match2(__attribute__((unused)) const char *pattern,
+                     __attribute__((unused)) const char *str)
+{
+    return 0;
+}
+
+OSListNode *__wrap_OSList_GetFirstNode(__attribute__((unused)) OSList *list)
+{
+    return NULL;
+}
+
+OSListNode *__wrap_OSList_GetNextNode(__attribute__((unused)) OSList *list)
+{
+    return NULL;
+}
+
+/* ===================================================================
+ * Tests for is_file()
+ * =================================================================== */
+
+static void test_is_file_null_input(void **state)
+{
+    (void)state;
+    assert_int_equal(is_file(NULL), 0);
+}
+
+/* A directory counts as found — wopendir succeeds */
+static void test_is_file_found_by_opendir(void **state)
+{
+    (void)state;
+    DIR *fake_dp = (DIR *)0x1234;
+
+    expect_string(__wrap_wopendir, name, "/bin/some_dir");
+    will_return(__wrap_wopendir, fake_dp);
+
+    assert_int_equal(is_file("/bin/some_dir"), 1);
+}
+
+/* wopendir fails, w_stat succeeds */
+static void test_is_file_found_by_stat(void **state)
+{
+    (void)state;
+
+    expect_string(__wrap_wopendir, name, "/bin/ps");
+    will_return(__wrap_wopendir, NULL);
+
+    expect_string(__wrap_w_stat, pathname, "/bin/ps");
+    will_return(__wrap_w_stat, 0);
+
+    assert_int_equal(is_file("/bin/ps"), 1);
+}
+
+/* wopendir fails, w_stat fails, waccess succeeds */
+static void test_is_file_found_by_access(void **state)
+{
+    (void)state;
+
+    expect_string(__wrap_wopendir, name, "/bin/ps");
+    will_return(__wrap_wopendir, NULL);
+
+    expect_string(__wrap_w_stat, pathname, "/bin/ps");
+    will_return(__wrap_w_stat, -1);
+
+    expect_string(__wrap_waccess, path, "/bin/ps");
+    will_return(__wrap_waccess, 0);
+
+    assert_int_equal(is_file("/bin/ps"), 1);
+}
+
+/* wopendir fails, w_stat fails, waccess fails, wfopen succeeds */
+static void test_is_file_found_by_fopen(void **state)
+{
+    (void)state;
+    FILE *fake_fp = (FILE *)0x5678;
+
+    expect_string(__wrap_wopendir, name, "/bin/ps");
+    will_return(__wrap_wopendir, NULL);
+
+    expect_string(__wrap_w_stat, pathname, "/bin/ps");
+    will_return(__wrap_w_stat, -1);
+
+    expect_string(__wrap_waccess, path, "/bin/ps");
+    will_return(__wrap_waccess, -1);
+
+    expect_string(__wrap_wfopen, path, "/bin/ps");
+    will_return(__wrap_wfopen, fake_fp);
+
+    assert_int_equal(is_file("/bin/ps"), 1);
+}
+
+/* All methods fail → not found */
+static void test_is_file_not_found(void **state)
+{
+    (void)state;
+
+    expect_string(__wrap_wopendir, name, "/bin/ps");
+    will_return(__wrap_wopendir, NULL);
+
+    expect_string(__wrap_w_stat, pathname, "/bin/ps");
+    will_return(__wrap_w_stat, -1);
+
+    expect_string(__wrap_waccess, path, "/bin/ps");
+    will_return(__wrap_waccess, -1);
+
+    expect_string(__wrap_wfopen, path, "/bin/ps");
+    will_return(__wrap_wfopen, NULL);
+
+    assert_int_equal(is_file("/bin/ps"), 0);
+}
+
+/* ===================================================================
+ * Main
+ * =================================================================== */
+
+int main(void)
+{
+    const struct CMUnitTest tests[] = {
+        cmocka_unit_test(test_is_file_null_input),
+        cmocka_unit_test(test_is_file_found_by_opendir),
+        cmocka_unit_test(test_is_file_found_by_stat),
+        cmocka_unit_test(test_is_file_found_by_access),
+        cmocka_unit_test(test_is_file_found_by_fopen),
+        cmocka_unit_test(test_is_file_not_found),
+    };
+
+    return cmocka_run_group_tests(tests, NULL, NULL);
+}

